### PR TITLE
For Pytorch2.13 ，Major Breakthroughs: Sequential, Module, ModuleList, ModuleDict, Embedding & EmbeddingBag add chain builder mode support more features

### PR DESCRIPTION
**Pull request by mullerhai**
PyTorch: richer nn.Module / Sequential API, package split, NCCL→distributed, ABI-safe header patches
Shorter alternative:

PyTorch: Module multi-forward, Sequential push_back, fluent Options, Embedding.from_pretrained, package layout
PR body (copy-paste)

### Motivation
The current PyTorch presets expose only a thin slice of the C++ frontend. Building real models from Java is painful: `Sequential` cannot host arbitrary `Module` subclasses cleanly, `Module` has no multi-arity `forward`, Options are not chainable, `Embedding` lacks `from_pretrained`, and packages are a flat dump of generated peers. This PR makes the Java API much closer to the LibTorch C++ / Python experience while keeping builds portable across macOS / Linux / Windows.
### Summary
Major usability and packaging upgrades for the **PyTorch 2.13** presets:
1. Multi-arity `Module.forward(...)` overloads (Tensor / tuple / attention / ASM)
2. `AnyModule` ↔ `Module` interop and typed `asXxx()` helpers
3. `SequentialImpl` can `push_back` any `Module` subclass; constructors accept `OrderedDict` / named modules
4. Fluent chainable setters on all `*Options` hyper-parameter classes
5. `EmbeddingImpl` / `EmbeddingBagImpl.from_pretrained(...)`
6. Package relocation (`nn`, `data`, `distributed`, `jit`, `optim`, …)
7. `ProcessGroupNCCL` generated into `org.bytedeco.pytorch.distributed`
8. ABI-safe LibTorch header patches (fixes Dropout-in-Sequential crash)
9. Do not ship preset C++ glue headers inside the Java JAR
Verified:
- `mvn -pl pytorch install -DskipTests`
- `mvn -pl pytorch install -DskipTests -Djavacpp.cppbuild.skip=true`
- Runtime: Sequential with `DropoutImpl` / `ReLUImpl` / `LinearImpl`; MHA `forward` / `forwardT_*`
#### 1. `Module` multi-arity `forward` overloads
- Public Java `forward(...)` covering common LibTorch arities:
  - 1–4 / 6 / 8 Tensor args
  - `output_size` / indices + `output_size`
  - RNN/LSTM-style tuple returns (`T_TensorTensor_T`, `T_TensorT_TensorTensor_T_T`)
  - MultiheadAttention-style 7-arg attention forward
- C++ side: virtual `forward_tensor*` / `forward_tuple*` trampolines on `torch::nn::Module` (see header patches).
- Java-side `@Virtual` only on the 1-arg path where needed for user subclasses; multi-arg `@Virtual`s are stripped post-parse to avoid wrong-arity vtable dispatch inside `Sequential`.
#### 2. `AnyModule` / `Module` conversion & containers
- Reliable `AnyModule` construction from concrete `*Impl` types.
- `ModuleListImpl` / `ModuleDictImpl` insert / get helpers with `ModuleAsHelper` so Java identity is preserved across native round-trips.
- Typed `asLinear()`, `asDropout()`, … cast helpers for modules retrieved as base `Module`.
#### 3. `SequentialImpl` enhancements
- `push_back` for arbitrary modules that inherit `Module` (typed templates for built-ins + generic `Module` path).
- Constructors from `OrderedDict<string, AnyModule>` and `OrderedDict<string, shared_ptr<Module>>` (lvalue + rvalue), matching C++ `Sequential` usage.
- End-to-end forward through heterogeneous stacks (Linear / ReLU / Dropout / Flatten / …).
#### 4. Fluent `*Options` chain setters
- Post-parse generator (`NNOptionsFluentSetterGenerator`) injects chainable setters on generated `*Options` classes.
- Example: `new LinearOptions(in, out).bias(false).device(device)` style usage from Java.
#### 5. `Embedding` / `EmbeddingBag` `from_pretrained`
- Header adapters + Java mappings for:
  - `EmbeddingImpl.from_pretrained(embeddings, options)`
  - `EmbeddingBagImpl.from_pretrained(embeddings, options)`
  - `EmbeddingBagImpl.forward_with_offsets(input, offsets)` convenience
#### 6. Package layout (post-parse relocation)
Generated peers are relocated into a clearer layout, e.g.:
| Area | Package |
|------|---------|
| Core tensors / c10 leftovers | `org.bytedeco.pytorch` |
| `torch::nn` modules | `org.bytedeco.pytorch.nn.modules` (+ `container`, `options`, `functions`) |
| Data / datasets / samplers | `org.bytedeco.pytorch.data.*` |
| Distributed (c10d) | `org.bytedeco.pytorch.distributed` |
| JIT / optim / serialize / profiler / … | matching subpackages |
Script: `scripts/relocate_packages.py` (runs in `process-sources`).
#### 7. NCCL `ProcessGroupNCCL` → `distributed`
- `torch_nccl` preset `target` is now `org.bytedeco.pytorch.distributed` so `ProcessGroupNCCL` (and related peers) live next to `ProcessGroup`, `ProcessGroupGloo`, `Store`, etc.
- `module-info.java` exports updated; glue still loads via `global.torch_nccl`.
#### 8. ABI-safe LibTorch header patches (critical)

9. JAR packaging
Exclude preset C++ glue headers from the main artifact:
org/bytedeco/pytorch/presets/**/*.h
keep include/ excluded as before
Headers are only needed at parse/build time, not at runtime.
10. Multi-return modules (MultiheadAttention, RNN/GRU/LSTM, AdaptiveLogSoftmax)
Java cannot overload forward with the same arity and a different return type than Module.forward(Tensor...).
Those C++ forwards are mapped as forwardT_* / forwardASMoutput (same pattern as Module’s tuple forwards).
Convenience forward(...) wrappers where arity does not clash (e.g. MHA 7-arg; LSTM Optional-hx).
Documented on the generated methods.
11. Debug toString() printers
Injected toString() on Module, Tensor, containers, optimizers, samplers, common losses (via ModulePrinter / TensorPrinter / …), closer to Python print(model).
Build / test

# full native + Java
mvn -pl pytorch install -DskipTests
# Java-only against already-built natives
mvn -pl pytorch install -DskipTests -Djavacpp.cppbuild.skip=true
Runtime checks (samples / small harnesses):

Sequential: Linear → ReLU → Dropout → Linear (previously crashed)
MHA forward (7-arg) and forwardT_TensorTensor_T (3-arg)

### CODE REVIEW

**Comment by mullerhai** [comment]
@saudet  please review and consider merge , this is really big gap for previous version

**Comment by mullerhai** [comment]
Hi， @saudet   ，I have fix all  I wanted feature ，then I want to bring  more features to  javacpp-pytorch 。I have updated javacpp-pytorch so that torch.tensor() now accepts multi-dimensional array arguments spanning 1 through 6 dimensions. All related bugs have been fixed, and new features aligned with Python PyTorch behavior have been added. My goal is to turn it into a fully-featured, versatile training framework that gains genuine traction and popularity.
I plan to integrate its upstream and downstream toolchains and third-party packages, including NumPy, Pickle, GGUF, SafeTensors, TensorBoard, Parquet, DataFrames, CSV, KV Cache, FSDP and DDP.
I intend to publish beta-version JARs to Maven Central first, bundled alongside JavaCPP, OpenBLAS and CUDA bindings. I hope you can merge all these supplementary components. Once complete, we will truly have an outstanding framework.
Thank you.

**Comment by mullerhai** [comment]
Hi @saudet  ，


I have published some javacpp packages to maven center repo， because javacpp-presets so many sub projects， I make one mini javacpp presets project  only  “ javacpp， cpython ，numpy， openblas， cuda， tensorrt ，tritonserver ，opencv ，ffmpeg ， pytorch ，pytorch-gpu and cuda-redist，cuda-redist-cublas，cuda-redist-cudnn，cuda-redist-cusolver，cuda-redist-cusparse，cuda-redist-npp，cuda-redist-nccl，cuda-redist-nvcomp ” ，
···xml


        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cublas</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cudnn</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cusolver</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cusparse</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-npp</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-nccl</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>

        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-nvcomp</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cublas</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cudnn</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cusolver</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-cusparse</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-npp</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-nccl</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>cuda-redist-nvcomp</artifactId>
            <version>13.3-9.24-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>tritonserver</artifactId>
            <version>2.70.0-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>

        </dependency>


        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>javacpp</artifactId>
            <version>1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>

        </dependency>
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>openblas</artifactId>
            <version>0.3.33-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>

        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>ffmpeg</artifactId>
            <version>8.1.1-1.5.14-beta-08</version>
            
            <classifier>linux-x86_64</classifier>
            <optional>true</optional>
        </dependency>
        
        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>opencv</artifactId>
            <version>4.13.0-1.5.14-beta-08</version>
            <classifier>linux-x86_64</classifier>
            
            <optional>true</optional>
        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>pytorch</artifactId>
            <version>2.13.0-1.5.14-beta-08</version>

            
            <optional>true</optional>
        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>pytorch</artifactId>
            <version>2.13.0-1.5.14-beta-08</version>

            <classifier>linux-x86_64</classifier>
            <optional>true</optional>
        </dependency>

        <dependency>
            <groupId>io.github.mullerhai</groupId>
            <artifactId>pytorch-platform-gpu</artifactId>
            <version>2.13.0-1.5.14-beta-08</version>
            

        </dependency>


···

**Comment by devjeonghwan** [comment]
I am not speaking for the maintainers, but I think this PR needs a full review from the beginning.

It has 908 commits and changes more than 11,000 files. It also includes many changes that are not related to the PyTorch work. Some changes affect the whole project, such as requiring Java 17, moving existing packages, and removing many presets.

I suggest closing this PR, creating a new branch from the latest upstream `master`, and sending the changes again as small and clear PRs. Changes to the Java version and package structure should be discussed separately first.

**Comment by mullerhai** [comment]
> I am not speaking for the maintainers, but I think this PR needs a full review from the beginning.
> 
> It has 908 commits and changes more than 11,000 files. It also includes many changes that are not related to the PyTorch work. Some changes affect the whole project, such as requiring Java 17, moving existing packages, and removing many presets.
> 
> I suggest closing this PR, creating a new branch from the latest upstream `master`, and sending the changes again as small and clear PRs. Changes to the Java version and package structure should be discussed separately first.

I think we need create new small mini javacpp presets ，only javacpp-pytorch and it tools chains dependencies，I have add a lot tools code make it fat ，but make it more like python pytorch useful and powerful ，how do you thinks
